### ISSUE TITLE
Allow grouping/headers only in recipe-wide ingredients list

### ISSUE THREAD

**Issue by JadElClemens** [vabene1111/recipes#944]
**Is your feature request related to a problem? Please describe.**
I'm copying over [this recipe for gyuudon](https://www.justonecookbook.com/yoshinoya-beef-bowl-gyudon/) (manually because of #170 but that's neither here nor there) and would like to create certain sub-groups of ingredients - specifically, I want a group for the sauce/soup base ingredients.

Warning: This is a little long-winded and may not be very clear, my apologies in advance. To clarify, I'm using "recipe-wide ingredients list" to refer to the combined list of all ingredients defined within the steps of the recipe.

I'm not a master of mise en place, and I feel like the design of Tandoor already enables me a bit by having me specify ingredients step-by-step so that I _can_ prep as I go along instead of doing all of my prep up-front - evidently this bit of the design is explicit per a comment by @vabene1111 on another issue I saw while checking if this had already been requested.

However, in this particular instance, the "sauce" ingredients are the one thing I _would_ want to prep beforehand, and to support this I'd like to add a header to the ingredients in the "make sauce" step (step 4 in the original recipe) that clarify - from the recipe-wide ingredients view - that these and only these ingredients are involved in the sauce and that I should prep those first. Well, to make the whole list look uniform as a personal preference, I also want to add headers categorizing the rest of the ingredients - perhaps something like "Main ingredients" and "Toppings" as well. Problem is, the ingredients that _would_ go into "Toppings" are split between steps and so I have to do this weird formatting where I add the "Toppings" header to the first step involving such an ingredient and leave it out of the second step, which feels a little strange.

I can only imagine this issue would be compounded if there were a recipe with ingredients which should be grouped together but are listed out of order.

**Describe the solution you'd like**
It would be nice if there were a way to re-arrange and add labels/headers to groups in the recipe-wide ingredients list. Maybe as you edit and update a recipe there could be a preview of the recipe-wide ingredients that can be re-arranged and have headers added to it.

That, or maybe ingredients could have a new (per-recipe) field added that specifies the recipe group to add it to within the recipe-wide ingredients list, then display these groups with a header in the recipe-wide list. From a programmatic standpoint that seems like a easier solution to me.

**Describe alternatives you've considered**
I considered and implemented the little hack I described and screenshotted below. Alternatively I would perhaps have to add an extra step up front to prepare the sauce mixture.

**Additional context**
My recipe ingredients list:
![image](https://user-images.githubusercontent.com/11086175/135696891-bfd44cb0-ceb0-4b4a-9865-760699f76d5f.png)

My hack to get the list looking like I want:
![image](https://user-images.githubusercontent.com/11086175/135696895-19bf0dc4-0ccc-4774-a4c2-e024bf105e7b.png)

**Comment by vabene1111** [comment]
Hmm ok so first of all thanks for the feedback. I am not 100% sure if I am getting your problem right.

Basically you want to have one step for every real world step (like chop onions, clean meat, peel potatoes) but you want those ingredients to be displayed under one combined header.

I don't really see what is wrong with the solution you showed that just put the header ingredient first in the first of those steps.

I usually combine those things into one step and just make a small list in the instructions.

Having a centralized list in addition to the step based list is, from a technical perspective, a little weird so I would try to not implement something like that.

**Comment by smilerz** [comment]
@JadElClemens is the picture you showed what the ingredient list looks like now?  Or how you'd like it to look?

Also, RE: #170 - the new Automation functionality will effectively fix that for you.  Import the unit (or food, or whatever) with the wrong case (TBSP), and create a merge/automation (TBSP -> tbsp) and it will automatically import as the correct the unit from that point forward.

**Comment by JadElClemens** [comment]
> Basically you want to have one step for every real world step (like chop onions, clean meat, peel potatoes) but you want those ingredients to be displayed under one combined header.

Not necessarily, I guess what I really want is the ability to assign ingredients to certain groups regardless of what steps they are part of.

> I don't really see what is wrong with the solution you showed that just put the header ingredient first in the first of those steps.

Nothing for this recipe, but I can imagine (in theory) a recipe that would not allow for grouping ingredients in this way. What if the "Sauce" ingredients were for some reason spread out between steps 1, 3, and 9 for example? Not that that's a realistic way to write _this_ recipe. Perhaps for this one I can see it being an issue if I had to prep a topping (green onions perhaps?) - I would in theory like to have that listed up top with the rest of the cutting but still be listed in the global list as a "topping".

> Having a centralized list in addition to the step based list is, from a technical perspective, a little weird so I would try to not implement something like that.
Sure, that was just one suggestion. I think the grouping concept makes more sense.

-----
> @JadElClemens is the picture you showed what the ingredient list looks like now? Or how you'd like it to look?
The first picture is both how I want it to and how it currently does look. The second image shows the hack I had to do with the ingredients lists in order to make it look how I wanted.

**Comment by vabene1111** [comment]
ok now i think i better unterstand what you mean. 
If both the lets say the dough, the sauce and the toppings require work spread out possibly over some hours/days (as is not atypical with dough especially) everything would be mixed in the central ingredient list. 

Although i understand the problem i am not sure if its big enough to 
a) justify spending quite some time finding a possible technical solution that still allows ingredients and steps to be linked (which is still required) 
and b) making the structure of a recipe even more complicated so that the average user who just uses a single step most of the time does not get overwhelmed by even more additional options. 

I think for now the solution would be to simply split the recipe in 3 one for each "component" lets say and cook them parallel to each other where #636 will help in the future. 

I am planning to look at how recipes are composed of steps and the issues resolving around this and will think about this issue as well when doing so.

**Comment by tomtjes** [comment]
Drafted a feature request, then found this one. So I'm just adding my request here, which I think summarizes the above discussion well and describes a solution that hopefully is not too complex.


## Assign dish components to recipe steps

Traditionally, recipe ingredients are often grouped by the various components of a dish:

```
Sauce:
- tomatoes
- anchovies
- salt

Topping: 
- parsley
- parmesan
```

This can be done in Tandoor by adding headers to the list of ingredients. However, this has its limitations when ingredients are attached to multiple steps. 

If step names are defined, they can be shown in the list of ingredients. But as recipes are written chronologically and preparation often alternates between various components (e.g. 1-prepare something for the sauce, 2-prepare main component, 3-return to working on the sauce), this has its limitations, too.

I would like to suggest a feature that lets you assign a dish component to a recipe step. Components could either be defined per recipe or globally. If components are assigned to one or more steps in a recipe, the ingredient list would no longer display step names and instead group the ingredients by component name. If some steps don't have a component assigned, those ingredients should appear at the top of the ingredient list.

```
Spaghetti puttanesca

Ingredients

- 500g spaghetti
- 1 tbsp salt

For the sauce:
- 1 can tomatoes
- 5 anchovy fillets
- 1 tsp salt

To serve:
- parmesan
```
In the above example, steps that use the ingredients `500g spaghetti` and `1 tbsp salt` don't have a component assigned to them. Steps that use the sauce ingredients have `For the sauce:` assigned as their component. These ingredients could be used in multiple steps throughout the recipes. Ideally, if one sauce step would use `1/2 tsp salt` and a later sauce step would also use `1/2 tsp salt`, it would be consolidated as `1 tsp salt` in the list of the sauce ingredients, but it would be kept separate from the salt used in other steps.

Separating step names and dish components in this way allows for getting a better sense of what comprises a dish just by taking a glimpse at the main ingredient list, as well as for a quicker sense of the preparation method, when more descriptive step names are used.
